ambry-clustermap/src/main/java/com.github.ambry.clustermap/DynamicClusterChangeHandler.java

```diff
@@ -0,0 +1,487 @@
+    }
+  }
+
+  /**
+   * If partition override is enabled, we override replica SEAL/UNSEAL state based on partitionOverrideMap. If disabled,
+   * update replica state according to the info from {@link InstanceConfig}.
+   * @param replica the {@link ReplicaId} whose states (seal,stop) should be updated.
+   * @param sealedReplicas a list of {@link ReplicaId}(s) that are in SEALED state.
+   * @param stoppedReplicas a list of {@link ReplicaId}(s) that are in STOPPED state.
+   */
+  private void updateReplicaStateAndOverrideIfNeeded(AmbryReplica replica, List<String> sealedReplicas,
+      List<String> stoppedReplicas) {
+    String partitionName = replica.getPartitionId().toPathString();
+    boolean isSealed;
+    if (clusterMapConfig.clusterMapEnablePartitionOverride && partitionOverrideInfoMap.containsKey(partitionName)) {
+      isSealed = partitionOverrideInfoMap.get(partitionName)
+          .get(ClusterMapUtils.PARTITION_STATE)
+          .equals(ClusterMapUtils.READ_ONLY_STR);
+    } else {
+      isSealed = sealedReplicas.contains(partitionName);
+    }
+    replica.setSealedState(isSealed);
+    replica.setStoppedState(stoppedReplicas.contains(partitionName));
+  }
+
+  /**
+   * Create a new instance(node) and initialize disks/replicas on it.
+   * @param instanceConfig the {@link InstanceConfig} to create new instance
+   * @throws Exception
+   */
+  private void createNewInstance(InstanceConfig instanceConfig) throws Exception {
+    String instanceName = instanceConfig.getInstanceName();
+    logger.info("Adding node {} and its disks and replicas", instanceName);
+    AmbryDataNode datanode =
+        new AmbryDataNode(getDcName(instanceConfig), clusterMapConfig, instanceConfig.getHostName(),
+            Integer.valueOf(instanceConfig.getPort()), getRackId(instanceConfig), getSslPortStr(instanceConfig),
+            getXid(instanceConfig), helixClusterManagerCallback);
+    // for new instance, we first set it to unavailable and rely on its participation to update its liveness
+    if (!instanceName.equals(selfInstanceName)) {
+      datanode.setState(HardwareState.UNAVAILABLE);
+    }
+    initializeDisksAndReplicasOnNode(datanode, instanceConfig);
+    instanceNameToAmbryDataNode.put(instanceName, datanode);
+  }
+
+  /**
+   * Initialize the disks and replicas on the given node. Create partitions if this is the first time a replica of
+   * that partition is being constructed. If partition override is enabled, the seal state of replica is determined by
+   * partition info in HelixPropertyStore, if disabled, the seal state is determined by instanceConfig.
+   * @param datanode the {@link AmbryDataNode} that is being initialized.
+   * @param instanceConfig the {@link InstanceConfig} associated with this datanode.
+   * @throws Exception if creation of {@link AmbryDisk} throws an Exception.
+   */
+  private void initializeDisksAndReplicasOnNode(AmbryDataNode datanode, InstanceConfig instanceConfig)
```

Comment:
this function is quite similar to the one in the SimpleClusterChangeHandler. If we are going to keep two implementations for a long time, we can find a way to reuse the code.

Reply:
Yes, both of cluster change handlers should be able to share the code. I will probably do this in future PR after dynamic handler gets rolled out to production.

Reply:
Could we remove SimpleClusterChangeHandler after we have gained enough confidence with DynamicClusterChangeHandler?